.cursor/rules/code-style.mdc

```diff
@@ -0,0 +1,469 @@
+           assert result == "processed: test content"
+       
+       def test_error_handling(self) -> None:
+           """Test error scenarios."""
+           self.os_adaptor.read_file.side_effect = FileNotFoundError()
+           with pytest.raises(FileNotFoundError):
+               self.example.process_file("nonexistent.txt")
+       
+       def test_edge_cases(self) -> None:
+           """Test boundary conditions."""
+           self.os_adaptor.read_file.return_value = ""
+           result = self.example.process_file("empty.txt")
+           assert result == "processed: "
+   ```
+
+### Test Quality Requirements
+
+**Note**: Configuration modules that only provide default parameters are tested in integration tests, not unit tests, to ensure default parameters work correctly in real usage scenarios.
+
+1. **Use descriptive test names**:
+   ```python
+   # ✅ Good
+   def test_process_file_returns_processed_content_when_file_exists(self) -> None:
+   
+   # ❌ Bad
+   def test_process_file(self) -> None:
+   ```
+
+2. **Test both success and failure cases**:
+   ```python
+   def test_success_scenario(self) -> None:
+       # Test normal operation
+       pass
+   
+   def test_failure_scenario(self) -> None:
+       # Test error handling
+       pass
+   
+   def test_edge_cases(self) -> None:
+       # Test boundary conditions
+       pass
+   ```
+
+3. **Use proper mocking**:
+   ```python
+   # ✅ Correct - mock adaptors
+   @patch('dd_license_attribution.adaptors.os.OSAdaptor')
+   def test_with_mocked_adaptor(self, mock_adaptor: Mock) -> None:
+       mock_adaptor.return_value.read_file.return_value = "test"
+       # Test implementation
+   
+   # ❌ Incorrect - mock internal implementation
+   @patch('dd_license_attribution.example.ExampleClass._internal_method')
+   def test_with_mocked_internal(self, mock_internal: Mock) -> None:
+       # Don't mock internal implementation details
+   ```
+
+### Pre-commit Checklist
+
+Before committing any code changes:
+- [ ] Run `mypy src/ tests/` and ensure no errors
+- [ ] Verify all new functions have complete type annotations
+- [ ] Check that all variables have appropriate type hints
+- [ ] Ensure no `# type: ignore` comments were added without justification
+- [ ] Verify no direct OS imports in `src/` code
+- [ ] Create contract tests for any new external library usage
+- [ ] Run `pytest --cov=src/dd_license_attribution --cov-fail-under=95`
+- [ ] Ensure all new code has corresponding unit tests
+- [ ] Update CHANGELOG.md with any user-facing changes
+- [ ] Run `isort --check-only src/ tests/` and `black --check src/ tests/`
+- [ ] Test that CI MyPy check passes
+- [ ] Test that CI coverage check passes
+- [ ] Test that CI formatting check passes
+
+### Exceptions and Edge Cases
+
+- Use `Any` sparingly and only when the type truly cannot be determined
+- Use `Union[Type1, Type2]` or `Type1 | Type2` (Python 3.10+) for multiple possible types
```

Comment:
do we have a preference for which option is the default when they can both be used?
Maybe whichever one is most prevalent in the codebase?

Reply:
Given we require 3.10+ now, we should limit to the second only, https://peps.python.org/pep-0604/#add-a-new-operator-for-union-type1-type2 is implemented already in 3.10.
Updating in next commit.